code/fmri/load_save.py

```diff
@@ -0,0 +1,367 @@
+# emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: nil -*-
+# vi: set ft=python sts=4 ts=4 sw=4 et:
+#
+# Copyright 2023 The Axon Lab <theaxonlab@gmail.com>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# We support and encourage derived works from this project, please read
+# about our expectations at
+#
+#     https://www.nipreps.org/community/licensing/
+#
+""" Python module for loading and saving fMRI related data"""
+
+import os
+import os.path as op
+from collections import defaultdict
+import logging
+from typing import Optional, Union
+
+from bids import BIDSLayout
+import numpy as np
+from nibabel import loadsave
+from bids.layout import parse_file_entities
+from bids.layout.writing import build_path
+from nilearn.datasets import fetch_atlas_difumo
+from nilearn.interfaces.fmriprep.load_confounds import _load_single_confounds_file
+
+
+TIMESERIES_PATTERN: list = [
+    "sub-{subject}[/ses-{session}]/func/sub-{subject}"
+    "[_ses-{session}][_task-{task}][_desc-{desc}]"
+    "_{suffix}{extension}"
+]
+TIMESERIES_FILLS: dict = {"desc": "denoised", "extension": ".tsv"}
+CONFOUND_PATTERN: list = [
+    "sub-{subject}[_ses-{session}][_task-{task}][_part-{part}][_desc-{desc}]"
+    "_{suffix}{extension}"
+]
+CONFOUND_FILLS: dict = {"desc": "confounds", "suffix": "timeseries", "extension": "tsv"}
+
+
+def separate_by_similar_values(
+    input_list: list, external_value: Optional[Union[list, np.ndarray]] = None
+) -> dict:
+    """This returns elements of `input_list` with similar values (optionally set by
+    `external_value`) separated into sub-lists.
+
+    Parameters
+    ----------
+    input_list : list
+        List to be separated.
+    external_value : Optional[list], optional
+        Values corresponding to the elements of `input_list`, by default None
+
+    Returns
+    -------
+    dict
+        Dictionnary where each entry is a list of elements that have similar values and
+        the keys are the value for each list.
+    """
+    if external_value is None:
+        external_value = input_list
+
+    data_by_value = defaultdict(list)
+
+    for val, data in zip(external_value, input_list):
+        data_by_value[val].append(data)
+    return data_by_value
+
+
+def get_func_filenames_bids(
+    paths_to_func_dir: str,
+    task_filter: list = [],
+    ses_filter: list = [],
+    run_filter: list = [],
```

Comment:
Python: never use lists as default value for a function argument because they are mutable. Either make the default be `None` or allow tuples and pass an empty tuple:
```suggestion
    task_filter: Optional[list] = None,
    ses_filter: Optional[list] = None,
    run_filter: Optional[list] = None,
```

cc/ @celprov - while this nitpick has zero impact on our purpose because no one will take advantage of the mutability of lists in this context, this is critical when you deploy applications that people use in the wild (e.g., within a company).